File: exemples/c#/EnemyDecorator.cs
using Interfaces;
namespace PatternDecorators {
    abstract class EnemyDecorator: IEnemy {
        protected IEnemy enemy;

        public EnemyDecorator (IEnemy enemy) {
            this.enemy = enemy;
        }

        public int life;

        public int takeDamage(){
            return this.enemy.takeDamage();
        }

        public int attackDamage() {
            return this.enemy.attackDamage();
        }

    }
}

File: exemples/TypeScrypt/EnemyDecorator.ts
import { Enemy } from "./interfaces/Enemy";

export default abstract class EnemyDecorator implements Enemy {
    protected enemy;
    
    constructor(enemy: Enemy) {
        this.enemy = enemy
    }
    
    life: number;

    takeDamage(): number {
        return this.takeDamage();
    }

    attackDamage(): number {
        return this.attackDamage();
    }
}